File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/render/NativeBuffer.java
package org.maplibre.nativeffi.render;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.lang.ref.Cleaner;

/** Explicit off-heap byte buffer for reusable native readback and upload storage. */
public final class NativeBuffer implements AutoCloseable {
  private static final Cleaner CLEANER = Cleaner.create();

  private final NativeReference nativeReference;
  private final Cleaner.Cleanable cleanable;
  private final MemorySegment segment;
  private final long byteLength;

  private boolean closed;

  private NativeBuffer(long byteLength) {
    if (byteLength < 0) {
      throw new IllegalArgumentException("byteLength must be non-negative");
    }
    var arena = Arena.ofShared();
    this.nativeReference = new NativeReference(arena);
    this.cleanable = CLEANER.register(this, nativeReference);
    this.byteLength = byteLength;
    this.segment = byteLength == 0 ? MemorySegment.NULL : arena.allocate(byteLength);
  }

  public static NativeBuffer allocate(long byteLength) {
    return new NativeBuffer(byteLength);
  }

  public synchronized long byteLength() {
    ensureOpen();
    return byteLength;
  }

  public synchronized byte[] toByteArray() {
    ensureOpen();
    var length = Math.toIntExact(byteLength);
    if (length == 0) {
      return new byte[0];
    }
    return segment.toArray(ValueLayout.JAVA_BYTE);
  }

  synchronized MemorySegment segment() {
    ensureOpen();
    return segment;
  }

  synchronized void ensureCapacity(long requiredBytes) {
    ensureOpen();
    if (byteLength < requiredBytes) {
      throw new IllegalArgumentException("buffer is smaller than required byte length");
    }
  }

  private void ensureOpen() {
    if (closed) {
      throw new IllegalStateException("NativeBuffer is already closed");
    }
  }

  @Override
  public synchronized void close() {
    if (closed) {
      return;
    }
    closed = true;
    nativeReference.close();
    cleanable.clean();
  }

  private static final class NativeReference implements Runnable {
    private final Arena arena;
    private boolean closed;

    NativeReference(Arena arena) {
      this.arena = arena;
    }

    synchronized void close() {
      if (!closed) {
        closed = true;
        arena.close();
      }
    }

    @Override
    public void run() {
      close();
    }
  }
}

File: bindings/kotlin/src/jvmMain/kotlin/org/maplibre/nativeffi/render/NativeBuffer.kt
package org.maplibre.nativeffi.render

import java.lang.foreign.Arena
import java.lang.foreign.MemorySegment
import java.lang.foreign.ValueLayout
import java.lang.ref.Cleaner
import org.maplibre.nativeffi.internal.lifecycle.BorrowedResourceCore

private val nativeBufferCleaner: Cleaner = Cleaner.create()

/** Explicit off-heap byte buffer for reusable native readback and upload storage. */
public actual class NativeBuffer
private constructor(private val segment: MemorySegment, private val length: Long, arena: Arena) :
  AutoCloseable {
  private val nativeReference = NativeReference(arena)
  private val cleanable = nativeBufferCleaner.register(this, nativeReference)
  private val core = BorrowedResourceCore("NativeBuffer") { releaseNative() }

  public actual fun byteLength(): Long = withOpenBuffer { length }

  public actual fun toByteArray(): ByteArray = withOpenBuffer {
    if (length == 0L) ByteArray(0) else segment.toArray(ValueLayout.JAVA_BYTE)
  }

  internal fun <T> borrow(block: (MemorySegment, Long) -> T): T = withOpenBuffer {
    block(segment, length)
  }

  internal fun ensureCapacity(requiredBytes: Long) {
    withOpenBuffer {
      require(length >= requiredBytes) { "buffer is smaller than required byte length" }
    }
  }

  public actual override fun close() = core.close()

  private fun <T> withOpenBuffer(block: () -> T): T = core.withOpenResource(block)

  private fun releaseNative() {
    nativeReference.close()
    cleanable.clean()
  }

  public actual companion object {
    public actual fun allocate(byteLength: Long): NativeBuffer {
      require(byteLength >= 0) { "byteLength must be non-negative" }
      val arena = Arena.ofShared()
      val segment = if (byteLength == 0L) MemorySegment.NULL else arena.allocate(byteLength)
      return NativeBuffer(segment, byteLength, arena)
    }
  }

  private class NativeReference(private val arena: Arena) : Runnable {
    @Volatile private var closed = false

    @Synchronized
    fun close() {
      if (!closed) {
        closed = true
        arena.close()
      }
    }

    override fun run() {
      close()
    }
  }
}
